modelgauge/prompt_pipeline.py

```diff
@@ -0,0 +1,152 @@
+import csv
+import sys
+import traceback
+from abc import abstractmethod, ABCMeta
+from collections import defaultdict
+from typing import Any, Dict
+
+from modelgauge.pipeline import Source, Pipe, Sink
+from modelgauge.prompt import TextPrompt
+from modelgauge.sut import PromptResponseSUT
+
+
+class PromptItem(dict):
+    def __init__(self, row):
+        super().__init__()
+        self.update(row)
+
+    def uid(self):
+        return self["UID"]
+
+    def prompt(self):
+        return self["Text"]
+
+    def __hash__(self):
+        return hash(self.uid())
+
+
+class PromptInput(metaclass=ABCMeta):
+    @abstractmethod
+    def __iter__(self):
+        pass
+
+    def __len__(self):
+        count = 0
+        for prompt in self:
+            count += 1
+        return count
+
+
+class CsvPromptInput(PromptInput):
+    def __init__(self, path):
+        super().__init__()
+        self.path = path
+
+    def __iter__(self):
+        with open(self.path, newline="") as f:
+            csvreader = csv.DictReader(f)
+            for row in csvreader:
+                yield PromptItem(row)
+
+
+class PromptOutput(metaclass=ABCMeta):
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        pass
+
+    @abstractmethod
+    def write(self, item, results):
+        pass
+
+
+class CsvPromptOutput(PromptOutput):
+    def __init__(self, path, suts):
+        super().__init__()
+        self.path = path
+        self.suts = suts
+        self.file = None
+        self.writer = None
+
+    def __enter__(self):
+        self.file = open(self.path, "w", newline="")
+        self.writer = csv.writer(self.file, quoting=csv.QUOTE_ALL)
+        headers = ["UID", "Text"]
+        self.writer.writerow(headers + [s for s in self.suts.keys()])
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.file.close()
+
+    def write(self, item: PromptItem, results):
+        row = [item.uid(), item.prompt()]
+        for k in self.suts:
+            if k in results:
+                row.append(results[k])
+            else:
+                row.append("")
+        self.writer.writerow(row)
+
+
+class PromptSource(Source):
+    def __init__(self, input: PromptInput):
+        super().__init__()
+        self.input = input
+
+    def new_item_iterable(self):
+        return self.input
+
+
+class PromptSutAssigner(Pipe):
+    def __init__(self, suts: dict[str, PromptResponseSUT]):
+        super().__init__()
+        self.suts = suts
+
+    def handle_item(self, item: PromptItem):
+        for sut_key in self.suts:
+            self.downstream_put((item, sut_key))
+
+
+class PromptSutWorkers(Pipe):
+    def __init__(self, suts: dict[str, PromptResponseSUT], workers=None):
+        if workers is None:
+            workers = 8
```

Comment:
Because Tyler had an issue if he didn't specify a --workers argument at the command line. That CLI option defaults to `None` currently, and so a `None` was getting passed in here, which was then blowing up, because passing in a `None` overrides whatever default param value was set.

I could have fixed it with more code in the CLI, but since I wanted these objects to be pretty friendly to novices, I figured I'd make this harder to use wrongly.